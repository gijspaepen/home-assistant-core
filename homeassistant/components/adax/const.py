"""Constants for the Adax integration."""

from typing import Final

from homeassistant.const import Platform

ACCOUNT_ID: Final = "account_id"
CLOUD = "Cloud"
CONNECTION_TYPE = "connection_type"
DOMAIN: Final = "adax"
LOCAL = "Local"
WIFI_SSID = "wifi_ssid"
WIFI_PSWD = "wifi_pswd"

PLATFORMS = [Platform.CLIMATE, Platform.SENSOR]
