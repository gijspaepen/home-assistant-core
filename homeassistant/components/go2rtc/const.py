"""Go2rtc constants."""

DOMAIN = "go2rtc"

CONF_DEBUG_UI = "debug_ui"
EXCLUSE_ERROR_MESSAGE = "Url and debug_ui cannot be set at the same time."
