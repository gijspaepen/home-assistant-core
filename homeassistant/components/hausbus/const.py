"""Constants for the Haus-Bus integration."""

from __future__ import annotations

import logging

LOGGER = logging.getLogger(__package__)

DOMAIN = "hausbus"
ATTR_ON_STATE = "on_state"
