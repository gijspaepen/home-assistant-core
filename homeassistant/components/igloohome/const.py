"""Constants for the igloohome integration."""

DOMAIN = "igloohome"

OAUTH2_AUTHORIZE = "https://auth.igloohome.co/login"
OAUTH2_TOKEN = "https://auth.igloohome.co/oauth2/token"
