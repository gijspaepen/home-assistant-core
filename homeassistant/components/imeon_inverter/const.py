"""Constant for Imeon component."""

DOMAIN = "imeon_inverter"
HUBNAME = "imeon_inverter_hub"
TIMEOUT = 10  # seconds
