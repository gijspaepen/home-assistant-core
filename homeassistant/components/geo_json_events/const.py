"""Define constants for the GeoJSON events integration."""

from __future__ import annotations

from typing import Final

from homeassistant.const import Platform

DOMAIN: Final = "geo_json_events"

PLATFORMS: Final = [Platform.GEO_LOCATION]

ATTR_EXTERNAL_ID: Final = "external_id"
DEFAULT_RADIUS_IN_KM: Final = 20.0
DEFAULT_RADIUS_IN_M: Final = 20000.0
DEFAULT_UPDATE_INTERVAL: Final = 300
SOURCE: Final = "geo_json_events"
CONF_UPDATE_INTERVAL: Final = "update_interval"

SIGNAL_DELETE_ENTITY: Final = "geo_json_events_delete_{}"
SIGNAL_UPDATE_ENTITY: Final = "geo_json_events_update_{}"
