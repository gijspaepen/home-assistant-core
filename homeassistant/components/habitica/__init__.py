"""The habitica integration."""

from http import HTTPStatus
import logging

from aiohttp import ClientResponseError
from habitipy.aio import HabitipyAsync
import voluptuous as vol

from homeassistant.config_entries import ConfigEntry
from homeassistant.const import (
    APPLICATION_NAME,
    ATTR_NAME,
    CONF_API_KEY,
    CONF_NAME,
    CONF_URL,
    CONF_VERIFY_SSL,
    Platform,
    __version__,
)
from homeassistant.core import HomeAssistant, ServiceCall
from homeassistant.exceptions import ConfigEntryNotReady
from homeassistant.helpers import config_validation as cv
from homeassistant.helpers.aiohttp_client import async_get_clientsession
from homeassistant.helpers.typing import ConfigType

from .const import (
    ATTR_ARGS,
    ATTR_DATA,
    ATTR_PATH,
    CONF_API_USER,
    DEVELOPER_ID,
    DOMAIN,
    EVENT_API_CALL_SUCCESS,
    SERVICE_API_CALL,
)
from .coordinator import HabiticaDataUpdateCoordinator
from .services import async_setup_services
from .types import HabiticaConfigEntry

_LOGGER = logging.getLogger(__name__)
CONFIG_SCHEMA = cv.config_entry_only_config_schema(DOMAIN)


PLATFORMS = [Platform.BUTTON, Platform.SENSOR, Platform.SWITCH, Platform.TODO]


SERVICE_API_CALL_SCHEMA = vol.Schema(
    {
        vol.Required(ATTR_NAME): str,
        vol.Required(ATTR_PATH): vol.All(cv.ensure_list, [str]),
        vol.Optional(ATTR_ARGS): dict,
    }
)


async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
    """Set up the Habitica service."""

    async_setup_services(hass)

    return True


async def async_setup_entry(
    hass: HomeAssistant, config_entry: HabiticaConfigEntry
) -> bool:
    """Set up habitica from a config entry."""

    class HAHabitipyAsync(HabitipyAsync):
        """Closure API class to hold session."""

        def __call__(self, **kwargs):
            return super().__call__(websession, **kwargs)

        def _make_headers(self) -> dict[str, str]:
            headers = super()._make_headers()
            headers.update(
                {"x-client": f"{DEVELOPER_ID} - {APPLICATION_NAME} {__version__}"}
            )
            return headers

    async def handle_api_call(call: ServiceCall) -> None:
        name = call.data[ATTR_NAME]
        path = call.data[ATTR_PATH]
        entries = hass.config_entries.async_entries(DOMAIN)

        api = None
        for entry in entries:
            if entry.data[CONF_NAME] == name:
                api = entry.runtime_data.api
                break
        if api is None:
            _LOGGER.error("API_CALL: User '%s' not configured", name)
            return
        try:
            for element in path:
                api = api[element]
        except KeyError:
            _LOGGER.error(
                "API_CALL: Path %s is invalid for API on '{%s}' element", path, element
            )
            return
        kwargs = call.data.get(ATTR_ARGS, {})
        data = await api(**kwargs)
        hass.bus.async_fire(
            EVENT_API_CALL_SUCCESS, {ATTR_NAME: name, ATTR_PATH: path, ATTR_DATA: data}
        )

    websession = async_get_clientsession(
        hass, verify_ssl=config_entry.data.get(CONF_VERIFY_SSL, True)
    )

    api = await hass.async_add_executor_job(
        HAHabitipyAsync,
        {
            "url": config_entry.data[CONF_URL],
            "login": config_entry.data[CONF_API_USER],
            "password": config_entry.data[CONF_API_KEY],
        },
    )
    try:
        user = await api.user.get(userFields="profile")
    except ClientResponseError as e:
        if e.status == HTTPStatus.TOO_MANY_REQUESTS:
            raise ConfigEntryNotReady(
                translation_domain=DOMAIN,
                translation_key="setup_rate_limit_exception",
            ) from e
        raise ConfigEntryNotReady(e) from e

    if not config_entry.data.get(CONF_NAME):
        name = user["profile"]["name"]
        hass.config_entries.async_update_entry(
            config_entry,
            data={**config_entry.data, CONF_NAME: name},
        )

    coordinator = HabiticaDataUpdateCoordinator(hass, api)
    await coordinator.async_config_entry_first_refresh()

    config_entry.runtime_data = coordinator
    await hass.config_entries.async_forward_entry_setups(config_entry, PLATFORMS)

    if not hass.services.has_service(DOMAIN, SERVICE_API_CALL):
        hass.services.async_register(
            DOMAIN, SERVICE_API_CALL, handle_api_call, schema=SERVICE_API_CALL_SCHEMA
        )

    return True


async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
    """Unload a config entry."""
    if len(hass.config_entries.async_entries(DOMAIN)) == 1:
        hass.services.async_remove(DOMAIN, SERVICE_API_CALL)
    return await hass.config_entries.async_unload_platforms(entry, PLATFORMS)
