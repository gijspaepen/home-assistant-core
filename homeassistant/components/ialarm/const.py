"""Constants for the iAlarm integration."""

from pyialarm import IAlarm

from homeassistant.components.alarm_control_panel import AlarmControlPanelEntityState

DATA_COORDINATOR = "ialarm"

DEFAULT_PORT = 18034

DOMAIN = "ialarm"

IALARM_TO_HASS = {
    IAlarm.ARMED_AWAY: AlarmControlPanelEntityState.ARMED_AWAY,
    IAlarm.ARMED_STAY: AlarmControlPanelEntityState.ARMED_HOME,
    IAlarm.DISARMED: AlarmControlPanelEntityState.DISARMED,
    IAlarm.TRIGGERED: AlarmControlPanelEntityState.TRIGGERED,
}
