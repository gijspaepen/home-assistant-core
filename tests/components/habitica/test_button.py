"""Tests for Habitica button platform."""

from collections.abc import Generator
from http import HTTPStatus
import re
from unittest.mock import patch

import pytest
from syrupy.assertion import SnapshotAssertion

from homeassistant.components.button import DOMAIN as BUTTON_DOMAIN, SERVICE_PRESS
from homeassistant.components.habitica.const import DEFAULT_URL, DOMAIN
from homeassistant.config_entries import ConfigEntryState
from homeassistant.const import ATTR_ENTITY_ID, Platform
from homeassistant.core import HomeAssistant
from homeassistant.exceptions import ServiceValidationError
from homeassistant.helpers import entity_registry as er

from .conftest import assert_mock_called_with

from tests.common import MockConfigEntry, load_json_object_fixture, snapshot_platform
from tests.test_util.aiohttp import AiohttpClientMocker


@pytest.fixture(autouse=True)
def button_only() -> Generator[None]:
    """Enable only the button platform."""
    with patch(
        "homeassistant.components.habitica.PLATFORMS",
        [Platform.BUTTON],
    ):
        yield


@pytest.mark.usefixtures("mock_habitica")
async def test_buttons(
    hass: HomeAssistant,
    config_entry: MockConfigEntry,
    snapshot: SnapshotAssertion,
    entity_registry: er.EntityRegistry,
) -> None:
    """Test button entities."""

    config_entry.add_to_hass(hass)
    await hass.config_entries.async_setup(config_entry.entry_id)
    await hass.async_block_till_done()

    assert config_entry.state is ConfigEntryState.LOADED

    await snapshot_platform(hass, entity_registry, snapshot, config_entry.entry_id)


@pytest.mark.parametrize(
    ("entity_id", "api_url"),
    [
        ("button.test_user_allocate_all_stat_points", "user/allocate-now"),
        ("button.test_user_buy_a_health_potion", "user/buy-health-potion"),
        ("button.test_user_revive_from_death", "user/revive"),
        ("button.test_user_start_my_day", "cron"),
    ],
)
async def test_button_press(
    hass: HomeAssistant,
    config_entry: MockConfigEntry,
    mock_habitica: AiohttpClientMocker,
    entity_id: str,
    api_url: str,
) -> None:
    """Test button press method."""

    config_entry.add_to_hass(hass)
    await hass.config_entries.async_setup(config_entry.entry_id)
    await hass.async_block_till_done()

    assert config_entry.state is ConfigEntryState.LOADED

    mock_habitica.post(f"{DEFAULT_URL}/api/v3/{api_url}", json={"data": None})

    await hass.services.async_call(
        BUTTON_DOMAIN,
        SERVICE_PRESS,
        {ATTR_ENTITY_ID: entity_id},
        blocking=True,
    )

    assert_mock_called_with(mock_habitica, "post", f"{DEFAULT_URL}/api/v3/{api_url}")


@pytest.mark.parametrize(
    ("entity_id", "api_url"),
    [
        ("button.test_user_allocate_all_stat_points", "user/allocate-now"),
        ("button.test_user_buy_a_health_potion", "user/buy-health-potion"),
        ("button.test_user_revive_from_death", "user/revive"),
        ("button.test_user_start_my_day", "cron"),
    ],
    ids=["allocate-points", "health-potion", "revive", "run-cron"],
)
@pytest.mark.parametrize(
    ("status_code", "msg"),
    [
        (HTTPStatus.TOO_MANY_REQUESTS, "Currently rate limited"),
        (HTTPStatus.BAD_REQUEST, "Unable to connect to Habitica"),
        (HTTPStatus.UNAUTHORIZED, "Unable to carry out this action"),
    ],
)
async def test_button_press_exceptions(
    hass: HomeAssistant,
    config_entry: MockConfigEntry,
    mock_habitica: AiohttpClientMocker,
    entity_id: str,
    api_url: str,
    status_code: HTTPStatus,
    msg: str,
) -> None:
    """Test button press exceptions."""

    config_entry.add_to_hass(hass)
    await hass.config_entries.async_setup(config_entry.entry_id)
    await hass.async_block_till_done()

    assert config_entry.state is ConfigEntryState.LOADED

    mock_habitica.post(
        f"{DEFAULT_URL}/api/v3/{api_url}",
        status=status_code,
        json={"data": None},
    )

    with pytest.raises(ServiceValidationError, match=msg):
        await hass.services.async_call(
            BUTTON_DOMAIN,
            SERVICE_PRESS,
            {ATTR_ENTITY_ID: entity_id},
            blocking=True,
        )

    assert_mock_called_with(mock_habitica, "post", f"{DEFAULT_URL}/api/v3/{api_url}")


async def test_button_unavailable(
    hass: HomeAssistant,
    config_entry: MockConfigEntry,
    aioclient_mock: AiohttpClientMocker,
    snapshot: SnapshotAssertion,
    entity_registry: er.EntityRegistry,
) -> None:
    """Test buttons are unavailable if conditions are not met."""

    aioclient_mock.get(
        f"{DEFAULT_URL}/api/v3/user",
        json=load_json_object_fixture("user_buttons_unavailable.json", DOMAIN),
    )
    aioclient_mock.get(re.compile(r".*"), json={"data": []})

    config_entry.add_to_hass(hass)
    await hass.config_entries.async_setup(config_entry.entry_id)
    await hass.async_block_till_done()

    assert config_entry.state is ConfigEntryState.LOADED

    await snapshot_platform(hass, entity_registry, snapshot, config_entry.entry_id)
