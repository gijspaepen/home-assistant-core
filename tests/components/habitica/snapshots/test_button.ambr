# serializer version: 1
# name: test_button_unavailable[button.test_user_allocate_all_stat_points-entry]
  EntityRegistryEntrySnapshot({
    'aliases': set({
    }),
    'area_id': None,
    'capabilities': None,
    'config_entry_id': <ANY>,
    'device_class': None,
    'device_id': <ANY>,
    'disabled_by': None,
    'domain': 'button',
    'entity_category': None,
    'entity_id': 'button.test_user_allocate_all_stat_points',
    'has_entity_name': True,
    'hidden_by': None,
    'icon': None,
    'id': <ANY>,
    'labels': set({
    }),
    'name': None,
    'options': dict({
    }),
    'original_device_class': None,
    'original_icon': None,
    'original_name': 'Allocate all stat points',
    'platform': 'habitica',
    'previous_unique_id': None,
    'supported_features': 0,
    'translation_key': <HabitipyButtonEntity.ALLOCATE_ALL_STAT_POINTS: 'allocate_all_stat_points'>,
    'unique_id': '00000000-0000-0000-0000-000000000000_allocate_all_stat_points',
    'unit_of_measurement': None,
  })
# ---
# name: test_button_unavailable[button.test_user_allocate_all_stat_points-state]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'friendly_name': 'test-user Allocate all stat points',
    }),
    'context': <ANY>,
    'entity_id': 'button.test_user_allocate_all_stat_points',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': 'unavailable',
  })
# ---
# name: test_button_unavailable[button.test_user_buy_a_health_potion-entry]
  EntityRegistryEntrySnapshot({
    'aliases': set({
    }),
    'area_id': None,
    'capabilities': None,
    'config_entry_id': <ANY>,
    'device_class': None,
    'device_id': <ANY>,
    'disabled_by': None,
    'domain': 'button',
    'entity_category': None,
    'entity_id': 'button.test_user_buy_a_health_potion',
    'has_entity_name': True,
    'hidden_by': None,
    'icon': None,
    'id': <ANY>,
    'labels': set({
    }),
    'name': None,
    'options': dict({
    }),
    'original_device_class': None,
    'original_icon': None,
    'original_name': 'Buy a health potion',
    'platform': 'habitica',
    'previous_unique_id': None,
    'supported_features': 0,
    'translation_key': <HabitipyButtonEntity.BUY_HEALTH_POTION: 'buy_health_potion'>,
    'unique_id': '00000000-0000-0000-0000-000000000000_buy_health_potion',
    'unit_of_measurement': None,
  })
# ---
# name: test_button_unavailable[button.test_user_buy_a_health_potion-state]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'friendly_name': 'test-user Buy a health potion',
    }),
    'context': <ANY>,
    'entity_id': 'button.test_user_buy_a_health_potion',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': 'unavailable',
  })
# ---
# name: test_button_unavailable[button.test_user_revive_from_death-entry]
  EntityRegistryEntrySnapshot({
    'aliases': set({
    }),
    'area_id': None,
    'capabilities': None,
    'config_entry_id': <ANY>,
    'device_class': None,
    'device_id': <ANY>,
    'disabled_by': None,
    'domain': 'button',
    'entity_category': None,
    'entity_id': 'button.test_user_revive_from_death',
    'has_entity_name': True,
    'hidden_by': None,
    'icon': None,
    'id': <ANY>,
    'labels': set({
    }),
    'name': None,
    'options': dict({
    }),
    'original_device_class': None,
    'original_icon': None,
    'original_name': 'Revive from death',
    'platform': 'habitica',
    'previous_unique_id': None,
    'supported_features': 0,
    'translation_key': <HabitipyButtonEntity.REVIVE: 'revive'>,
    'unique_id': '00000000-0000-0000-0000-000000000000_revive',
    'unit_of_measurement': None,
  })
# ---
# name: test_button_unavailable[button.test_user_revive_from_death-state]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'friendly_name': 'test-user Revive from death',
    }),
    'context': <ANY>,
    'entity_id': 'button.test_user_revive_from_death',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': 'unavailable',
  })
# ---
# name: test_button_unavailable[button.test_user_start_my_day-entry]
  EntityRegistryEntrySnapshot({
    'aliases': set({
    }),
    'area_id': None,
    'capabilities': None,
    'config_entry_id': <ANY>,
    'device_class': None,
    'device_id': <ANY>,
    'disabled_by': None,
    'domain': 'button',
    'entity_category': None,
    'entity_id': 'button.test_user_start_my_day',
    'has_entity_name': True,
    'hidden_by': None,
    'icon': None,
    'id': <ANY>,
    'labels': set({
    }),
    'name': None,
    'options': dict({
    }),
    'original_device_class': None,
    'original_icon': None,
    'original_name': 'Start my day',
    'platform': 'habitica',
    'previous_unique_id': None,
    'supported_features': 0,
    'translation_key': <HabitipyButtonEntity.RUN_CRON: 'run_cron'>,
    'unique_id': '00000000-0000-0000-0000-000000000000_run_cron',
    'unit_of_measurement': None,
  })
# ---
# name: test_button_unavailable[button.test_user_start_my_day-state]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'friendly_name': 'test-user Start my day',
    }),
    'context': <ANY>,
    'entity_id': 'button.test_user_start_my_day',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': 'unavailable',
  })
# ---
# name: test_buttons[button.test_user_allocate_all_stat_points-entry]
  EntityRegistryEntrySnapshot({
    'aliases': set({
    }),
    'area_id': None,
    'capabilities': None,
    'config_entry_id': <ANY>,
    'device_class': None,
    'device_id': <ANY>,
    'disabled_by': None,
    'domain': 'button',
    'entity_category': None,
    'entity_id': 'button.test_user_allocate_all_stat_points',
    'has_entity_name': True,
    'hidden_by': None,
    'icon': None,
    'id': <ANY>,
    'labels': set({
    }),
    'name': None,
    'options': dict({
    }),
    'original_device_class': None,
    'original_icon': None,
    'original_name': 'Allocate all stat points',
    'platform': 'habitica',
    'previous_unique_id': None,
    'supported_features': 0,
    'translation_key': <HabitipyButtonEntity.ALLOCATE_ALL_STAT_POINTS: 'allocate_all_stat_points'>,
    'unique_id': '00000000-0000-0000-0000-000000000000_allocate_all_stat_points',
    'unit_of_measurement': None,
  })
# ---
# name: test_buttons[button.test_user_allocate_all_stat_points-state]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'friendly_name': 'test-user Allocate all stat points',
    }),
    'context': <ANY>,
    'entity_id': 'button.test_user_allocate_all_stat_points',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': 'unknown',
  })
# ---
# name: test_buttons[button.test_user_buy_a_health_potion-entry]
  EntityRegistryEntrySnapshot({
    'aliases': set({
    }),
    'area_id': None,
    'capabilities': None,
    'config_entry_id': <ANY>,
    'device_class': None,
    'device_id': <ANY>,
    'disabled_by': None,
    'domain': 'button',
    'entity_category': None,
    'entity_id': 'button.test_user_buy_a_health_potion',
    'has_entity_name': True,
    'hidden_by': None,
    'icon': None,
    'id': <ANY>,
    'labels': set({
    }),
    'name': None,
    'options': dict({
    }),
    'original_device_class': None,
    'original_icon': None,
    'original_name': 'Buy a health potion',
    'platform': 'habitica',
    'previous_unique_id': None,
    'supported_features': 0,
    'translation_key': <HabitipyButtonEntity.BUY_HEALTH_POTION: 'buy_health_potion'>,
    'unique_id': '00000000-0000-0000-0000-000000000000_buy_health_potion',
    'unit_of_measurement': None,
  })
# ---
# name: test_buttons[button.test_user_buy_a_health_potion-state]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'friendly_name': 'test-user Buy a health potion',
    }),
    'context': <ANY>,
    'entity_id': 'button.test_user_buy_a_health_potion',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': 'unknown',
  })
# ---
# name: test_buttons[button.test_user_revive_from_death-entry]
  EntityRegistryEntrySnapshot({
    'aliases': set({
    }),
    'area_id': None,
    'capabilities': None,
    'config_entry_id': <ANY>,
    'device_class': None,
    'device_id': <ANY>,
    'disabled_by': None,
    'domain': 'button',
    'entity_category': None,
    'entity_id': 'button.test_user_revive_from_death',
    'has_entity_name': True,
    'hidden_by': None,
    'icon': None,
    'id': <ANY>,
    'labels': set({
    }),
    'name': None,
    'options': dict({
    }),
    'original_device_class': None,
    'original_icon': None,
    'original_name': 'Revive from death',
    'platform': 'habitica',
    'previous_unique_id': None,
    'supported_features': 0,
    'translation_key': <HabitipyButtonEntity.REVIVE: 'revive'>,
    'unique_id': '00000000-0000-0000-0000-000000000000_revive',
    'unit_of_measurement': None,
  })
# ---
# name: test_buttons[button.test_user_revive_from_death-state]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'friendly_name': 'test-user Revive from death',
    }),
    'context': <ANY>,
    'entity_id': 'button.test_user_revive_from_death',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': 'unknown',
  })
# ---
# name: test_buttons[button.test_user_start_my_day-entry]
  EntityRegistryEntrySnapshot({
    'aliases': set({
    }),
    'area_id': None,
    'capabilities': None,
    'config_entry_id': <ANY>,
    'device_class': None,
    'device_id': <ANY>,
    'disabled_by': None,
    'domain': 'button',
    'entity_category': None,
    'entity_id': 'button.test_user_start_my_day',
    'has_entity_name': True,
    'hidden_by': None,
    'icon': None,
    'id': <ANY>,
    'labels': set({
    }),
    'name': None,
    'options': dict({
    }),
    'original_device_class': None,
    'original_icon': None,
    'original_name': 'Start my day',
    'platform': 'habitica',
    'previous_unique_id': None,
    'supported_features': 0,
    'translation_key': <HabitipyButtonEntity.RUN_CRON: 'run_cron'>,
    'unique_id': '00000000-0000-0000-0000-000000000000_run_cron',
    'unit_of_measurement': None,
  })
# ---
# name: test_buttons[button.test_user_start_my_day-state]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'friendly_name': 'test-user Start my day',
    }),
    'context': <ANY>,
    'entity_id': 'button.test_user_start_my_day',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': 'unknown',
  })
# ---
